server/lib/src/idm/application.rs

```diff
@@ -0,0 +1,631 @@
+            return Err(OperationError::SessionExpired);
+        }
+
+        let account = Account::try_from_entry_ro(&entry, &mut self.qs_read).map_err(|e| {
+            admin_error!("Failed to search account {:?}", e);
+            e
+        })?;
+
+        if account.is_anonymous() {
+            return Err(OperationError::InvalidUuid);
+        }
+
+        let application = self
+            .apprs
+            .inner
+            .rs_set
+            .get(&lae.application)
+            .ok_or_else(|| {
+                admin_info!("Application {:?} not found", lae.application);
+                OperationError::NoMatchingEntries
+            })?;
+
+        // Check linked group membership
+        let linked_group_uuid = match application.linked_group {
+            Some(u) => u,
+            None => {
+                admin_warn!(
+                    "Application {:?} does not have a linked group.",
+                    application.name
+                );
+                return Ok(None);
+            }
+        };
+
+        // TODO: Should we check if the group still exists or just trust in ref. integrity?
+
+        if !account.groups.iter().any(|e| e.uuid() == linked_group_uuid) {
+            trace!(
+                "User {:?} not member of application {:?} linked group {:?}",
+                account.uuid,
+                application.uuid,
+                linked_group_uuid
+            );
+            return Ok(None);
+        }
+
+        match account.verify_application_password(&application, lae.cleartext.as_str())? {
+            Some(_) => {
+                let session_id = Uuid::new_v4();
+                security_info!(
+                    "Starting session {} for {} {}",
+                    session_id,
+                    account.spn,
+                    account.uuid
+                );
+
+                Ok(Some(LdapBoundToken {
+                    spn: account.spn,
+                    session_id,
+                    effective_session: LdapSession::UnixBind(account.uuid),
+                }))
+            }
+            None => {
+                security_info!("Account does not have a configured application password.");
+                Ok(None)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::event::CreateEvent;
+    use crate::idm::account::Account;
+    use crate::idm::application::Application;
+    use crate::idm::event::GenerateApplicationPasswordEvent;
+    use crate::idm::server::IdmServerTransaction;
+    use crate::idm::serviceaccount::{DestroyApiTokenEvent, GenerateApiTokenEvent};
+    use crate::prelude::*;
+    use compact_jwt::{JwsCompact, JwsEs256Verifier, JwsVerifier};
+    use kanidm_proto::v1::ApiToken;
+    use std::str::FromStr;
+    use std::time::Duration;
+
+    const TEST_CURRENT_TIME: u64 = 6000;
+
+    #[idm_test]
+    async fn test_idm_application_excludes(idms: &IdmServer, _idms_delayed: &mut IdmServerDelayed) {
```

Comment:
In general, more comments to the logic of what and why you are testing things is needed. See https://github.com/kanidm/kanidm/blob/master/server/lib/src/repl/tests.rs#L196 for example.

Reply:
Ok, comments added.